server/src/main/java/org/elasticsearch/cluster/metadata/MetadataMappingService.java

```diff
@@ -360,4 +368,87 @@ public TimeValue ackTimeout() {
                     }
                 });
     }
+
+    private Map<String, Object> populateColumnPositions(CompressedXContent mappingUpdateSource,
+                                                        CompressedXContent existingSource,
+                                                        Index index,
+                                                        ClusterState currentState) {
+        Map<String, Object> updatedSourceMap = XContentHelper.convertToMap(mappingUpdateSource.compressedReference(), true).v2();
+        Map<String, Object> existingSourceMap = null;
```

Comment:
I wonder if this is the right place to do this. Doesn't `mapperService.merge(updatedSourceMap, MergeReason.MAPPING_UPDATE);` otherwise take care of the mapipng update (and potentially already decompress the source)?

Would be good if we can avoid several map conversion roundtrips. Cluster state updates are performance critical

Reply:
Avoiding compressing/decompressing worked for unpartitioned tables only. For partitioned tables, decompressing was inevitable.